pkg/agent/endpoints/sds/handler.go

```diff
@@ -99,43 +99,61 @@ func (h *Handler) StreamSecrets(stream discovery_v2.SecretDiscoveryService_Strea
 			}).Debug("Received StreamSecrets request")
 			h.triggerReceivedHook()
 
-			// The nonce should be empty (if we've never sent a response) or
-			// match the last sent nonce, otherwise the request should be
-			// ignored.
-			if lastNonce != newReq.ResponseNonce {
-				h.c.Log.Warnf("Received unexpected nonce %q (expected %q); ignoring request", newReq.ResponseNonce, lastNonce)
-				continue
+			// If there's error detail, always log it
+			if newReq.ErrorDetail != nil {
+				h.c.Log.WithFields(logrus.Fields{
+					"resource_names": newReq.ResourceNames,
+					"error_detail":   newReq.ErrorDetail.Message,
+				}).Error("Envoy reported errors applying secrets")
 			}
 
-			req = newReq
+			// If we've previously sent a nonce, this must be a reply
+			if lastNonce != "" {
+
+				// The nonce should match the last sent nonce, otherwise
+				// it's stale and the request should be ignored.
+				if lastNonce != newReq.ResponseNonce {
+					h.c.Log.Warnf("Received unexpected nonce %q (expected %q); ignoring request", newReq.ResponseNonce, lastNonce)
+					continue
+				}
+
+				if newReq.VersionInfo == "" || newReq.VersionInfo != versionInfo {
+					// The caller has failed to apply the secrets. Wait until the
+					// next update to send down new info.
```

Comment:
will this request potentially contain new ResourceNames to subscribe to? if so, we're gonna ignore them...

Reply:
Well spotted, according to the XDS spec:
> Envoy may update the list of resource_names it presents to the management server in each DiscoveryRequest that ACK/NACKs a specific DiscoveryResponse.

So yes, this could indeed ignore a legitimate update of resource names.